ros_controllers/test/test_diff_drive_controller.cpp

```diff
@@ -0,0 +1,270 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "controller_parameter_server/parameter_server.hpp"
+#include "hardware_interface/robot_hardware.hpp"
+#include "lifecycle_msgs/msg/state.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rcutils/get_env.h"
+#include "ros_controllers/diff_drive_controller.hpp"
+#include "std_msgs/msg/string.hpp"
+#include "test_robot_hardware/test_robot_hardware.hpp"
+
+#include "gtest/gtest.h"
+#include <array>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+
+using lifecycle_msgs::msg::State;
+
+void spin(rclcpp::executors::MultiThreadedExecutor* exe)
+{
+   exe->spin();
+}
+
+class TestDiffDriveController : public ::testing::Test
+{
+ protected:
+   static void SetUpTestCase()
+   {
+      rclcpp::init(0, nullptr);
+   }
+
+   void SetUp()
+   {
+      test_robot = std::make_shared<test_robot_hardware::TestRobotHardware>();
+      test_robot->init();
+      left_wheel_names = {{test_robot->joint_name1, test_robot->joint_name2, test_robot->joint_name3}};
+      right_wheel_names = {{test_robot->joint_name1, test_robot->joint_name2, test_robot->joint_name3}};
+      op_mode = {{test_robot->write_op_handle_name1}};
+
+      pub_node = std::make_shared<rclcpp::Node>("velocity_publisher");
+      velocity_publisher = pub_node->create_publisher<geometry_msgs::msg::Twist>(controller_name + "/cmd_vel",
+                                                                                 rclcpp::SystemDefaultsQoS());
+   }
+
+   static void TearDownTestCase()
+   {
+      rclcpp::shutdown();
+   }
+
+   /// Publish velocity msgs
+   /**
+    *  time_from_start - delay between each points
+    *  points - vector of trajectories. Each velocity consists of 3 joints
+    */
+   void publish(double linear, double angular)
+   {
+      int wait_count = 0;
+      auto topic = velocity_publisher->get_topic_name();
+      while (pub_node->count_subscribers(topic) == 0) {
+         if (wait_count >= 5) {
+            auto error_msg = std::string("publishing to ") + topic + " but no node subscribes to it";
+            throw std::runtime_error(error_msg);
+         }
+         std::this_thread::sleep_for(std::chrono::milliseconds(100));
+         ++wait_count;
+      }
```

Comment:
This should only publish, why do we need this while loop?

Reply:
Nothing yet. Left it there since it came from the trajectory controller and this controller will likely publish odometry.